pkg/collectors/log_follower.go

```diff
@@ -0,0 +1,247 @@
+	splitLine := strings.SplitN(line, " ", 2) //nolint:gomnd // moving this to a var would make the code less clear
+	if len(splitLine) < 2 {                   //nolint:gomnd // moving this to a var would make the code less clear
+		return
+	}
+	timestampPart := splitLine[0]
+	lineContent := splitLine[1]
+	timestamp, err := time.Parse(time.RFC3339, timestampPart)
+	if err != nil {
+		// This is not a value line something went wrong
+		return
+	}
+
+	if seen, ok := logs.seenLines[timestamp]; ok {
+		for _, seenContent := range seen {
+			if seenContent == lineContent {
+				// Already have this line skip it
+				return
+			}
+		}
+	}
+	logs.seenLines[timestamp] = append(logs.seenLines[timestamp], lineContent)
+	logs.pruned = false
+	if logs.withTimeStamps {
+		_, err = writer.WriteString(line + "\n")
+	} else {
+		_, err = writer.WriteString(lineContent + "\n")
+	}
+	if err != nil {
+		log.Error(fmt.Errorf("failed to write log output to file"))
+	}
+}
+
+func (logs *LogsCollector) writeToLogFile() {
+	logs.wg.Add(1)
+	defer logs.wg.Done()
+
+	fileHandle, err := os.OpenFile(logs.logsOutputFileName, os.O_CREATE|os.O_WRONLY, logFilePermissions)
+	utils.IfErrorExitOrPanic(err)
+	defer fileHandle.Close()
+
+	for {
+		select {
+		case <-logs.quit:
+			// Consume the rest of the lines so we don't miss lines
+			for len(logs.lines) > 0 {
+				line := <-logs.lines
+				logs.consumeLine(line, fileHandle)
+			}
+			return
+		case line := <-logs.lines:
+			logs.consumeLine(line, fileHandle)
+		default:
+			if !logs.pruned {
+				log.Info("pruning seen lines")
+				for key := range logs.seenLines {
+					if time.Since(key) > keepDuration {
+						delete(logs.seenLines, key)
+					}
+				}
+				logs.pruned = true
+			} else {
+				time.Sleep(time.Microsecond)
+			}
+		}
+	}
+}
+
+func (logs *LogsCollector) processLines(line string) string {
+	if strings.ContainsRune(line, lineDelim) {
+		lines := strings.Split(line, "\n")
+		for n := 0; n < len(lines)-2; n++ {
+			logs.lines <- lines[n]
+		}
+		line = lines[len(lines)-1]
+	}
+	return line
+}
+
+func (logs *LogsCollector) poll() error {
+	logs.lpLock.RLock()
+	defer logs.lpLock.RUnlock()
+	sinceSeconds := int64(time.Since(logs.lastPoll).Seconds())
+	podLogOptions := v1.PodLogOptions{
+		SinceSeconds: &sinceSeconds,
+		Container:    contexts.PTPContainer,
+		Follow:       true,
+		Previous:     false,
+		Timestamps:   true,
+	}
+	podName, err := logs.client.FindPodNameFromPrefix(contexts.PTPNamespace, contexts.PTPPodNamePrefix)
+	if err != nil {
+		return fmt.Errorf("failed to poll: %w", err)
+	}
+	logs.SetLastPool()
+	podLogRequest := logs.client.K8sClient.CoreV1().
+		Pods(contexts.PTPNamespace).
+		GetLogs(podName, &podLogOptions).
+		Timeout(followTimeout)
+	stream, err := podLogRequest.Stream(context.TODO())
+	if err != nil {
+		return fmt.Errorf("failed to poll: %w", err)
+	}
+	defer stream.Close()
+
+	line := ""
+	start := time.Now()
```

Comment:
Is this necessary? I would suggest using `logs.lastPoll` as you're already getting it.

I suggest also moving the `logs.SetLastPoll` call to this point (after the `podLogRequest.Stream()`) as well since it would then not move the `logs.lastPoll` pointer forward if an error occurs in the `podLogsRequest.Stream()` call. And future log polls would attempt to use a longer `SinceSeconds` to make up any lost time.

Reply:
it might be even better to move `logs.SetLastPoll` to the *end* of the function and set it to `start` then. I'm not certain.

Reply:
I agree that last poll should replace start I made `SetLastPoll` return last poll. It shouldn't be at the end otherwise we wouldn't have overlapping coverage. 

We also need to adjust the keep time accordingly in case we have enough missed polls.

Reply:
setting of the the last poll is now done at the end of the function